File: next/dweb-browser/Helper/src/AdapterManager.cs
namespace DwebBrowser.Helper;

public class AdapterManager<T> where T: notnull
{
	private readonly Dictionary<T, int> _adapterOrderMap = new();
	private IEnumerable<T> _orderAdapters = new List<T>();

	public IEnumerable<T> Adapters { get => _orderAdapters; }

	/// <summary>
	///	
	/// </summary>
	/// <param name="adapter"></param>
	/// <param name="order"></param> 越大排序越靠后
	/// <returns></returns>
	public Func<bool> Append(T adapter, int order = 0)
	{
		_adapterOrderMap[adapter] = order;

		_orderAdapters = from entry in _adapterOrderMap
						 orderby entry.Value ascending
						 select entry.Key;

		return () => Remove(adapter);
	}

	public bool Remove(T adapter) => _adapterOrderMap.Remove(adapter);
}

File: android/MicroService/src/main/java/org/dweb_browser/microservice/help/AdapterManager.kt
package org.dweb_browser.microservice.help

open class AdapterManager<T> {
  private val adapterOrderMap = mutableMapOf<T, Int>()
  private var orderAdapters = listOf<T>()
  val adapters get() = orderAdapters
  fun append(order: Int = 0, adapter: T): () -> Boolean {
    adapterOrderMap[adapter] = order
    orderAdapters =
      adapterOrderMap.toList().sortedBy { (_, b) -> b }.map { (adapter) -> adapter }
    return { remove(adapter) }
  }

  fun remove(adapter: T) = adapterOrderMap.remove(adapter) != null
}